### DISCUSSION TITLE
oo/polymorphism

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#146] [Announcements]
# oo/polymorphism

Java程序员进阶之路，小白的零基础Java教程，认真聊聊 Java的三大特征：多态

https://tobebetterjavaer.com/oo/polymorphism.html

**Comment by duy1314** [comment]
清单3-1，为什么父类的构造方法在调用子类的实例方法时，父类并不清楚子类对象中的属性值是什么？(；′⌒`)

**Reply by itwanger** [reply]
这是因为构造方法是在对象创建的过程中被调用的，而子类对象的实例变量在父类的构造方法中还没有被初始化。在调用子类实例方法时，子类的实例变量已经被初始化了，因此父类构造方法并不能知道子类实例变量的值是什么。如果父类在构造方法中需要使用子类的实例变量，可以使用父类的构造方法传入参数的方式来初始化。

**Reply by duy1314** [reply]
让我困扰的是父类构造器中已经能调用到子类的方法了，但是却不知道子类的实例变量。哈哈

**Reply by dark-eternity** [reply]
我不理解的是，新增子对象时，会先调父类构造，而父类构造里竟然能调用被重写的非静态方法？非静态方法不是只有创建了对象后才能调用么，此时还是在执行父类构造，子类对象应该还不存在才对，这是为什么？对象到底是怎么创建的，我晕了

**Reply by dark-eternity** [reply]
希望作者看到回复下，有答案的话，感谢🙇‍
